### ISSUE TITLE
Adding Small Epsilon in linalg_eig_backward to Improve Numerical Stability on GPU

### ISSUE THREAD

**Issue by alexanderlerner** [pytorch/pytorch#147544]
### 🚀 The feature, motivation and pitch

Hi PyTorch Team,

My team and I work on physics-inspired ML models where we use torch.linalg.eigh to get the eigenvector corresponding to the lowest eigenvalue of a Hermitian matrix. We sometimes run into numerical issues during backpropagation as we repeat training iterations over time. Specifically we see:

`linalg_eigh_backward: The eigenvectors in the complex case are specified up to multiplication by e^{i phi}. The specified loss function depends on this quantity, so it is ill-defined`

Upon investigation, there are not actually phase dependencies in our case. After digging into `linalg_eig_backward`, we found the issue actually might come from the backward formula that divides by `lambda_j - lambda_i` when computing `Econj` [here](https://github.com/pytorch/pytorch/blob/382fbcc1e43ae5d46ec148bdfd8dcfb73da81b77/torch/csrc/autograd/FunctionsManual.cpp#L3757C1-L3765C1). For near-degenerate eigenvalues, we suspect the small denominator might trigger unstable gradients or NaNs in lower-precision (e.g., `complex64`) calculations as we accumulate these small errors over time.

Our fix has been to add a tiny epsilon in the denominator, something like:
`const float eps = 1e-8;
auto ret = VhgV.div_(Econj + eps);`

We found that this small regularizer effectively removes NaNs and stabilizes training without altering our models (which are just subclasses of `nn.module`) . 

We would like to potentially contribute a PR to PyTorch that conditionally adds this epsilon—potentially factoring in tensor precision (e.g., using a smaller epsilon for double precision).

Note that this only occurs on GPU, I'm running torch version `2.6.0` and cuda `12.3`. On CPU, the behavior shown actually throws a different error which makes more sense:

`torch._C._LinAlgError: linalg.eigh: (Batch element 0): The algorithm failed to converge because the input matrix is ill-conditioned or has too many repeated eigenvalues (error code: 7).`

Adding this epsilon in either case seemed to fix these issues -- some of our models are proprietary and we can't quite share an example, but might anyone know why we see this divergence in behavior between GPU and CPU, or at least what sorts of regimes might create this phenomenon in general?

We’re new to contributing to PyTorch and would appreciate any guidance. Thank you so much for your time, and we look forward to your feedback!

### Additional context

Some past issues referencing this same check:

https://github.com/pytorch/pytorch/pull/70528

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan @jianyuh @mruberry @walterddr @xwang233 @Lezcano

**Comment by lezcano** [comment]
I don't think I understand. The computation `VhgV.div_(Econj)` is *after* the check, so how does that help with the error?

Also, note that the error is triggered when the relative difference is greater than 0.01 (and same for the absolute difference) so it's quite generous. Perhaps you want to compute the eigendecomposition in `complex128` in this case to avoid all these numerical errors?

**Comment by alexanderlerner** [comment]
Hi @lezcano , thank you for your response! Apologies for the delayed reply as I wanted to investigate this further before responding.

From what I understand, It is after the check, but in our case the division of `VhgV.div_(Econj)` produces NaNs for the gradient because `Econj` is close to or equal to `0`. Then, in the next iteration, the NaNs that are produced seem to trigger the `TORCH_CHECK` in the subsequent iteration. 

We find that adding the `eps` to the Econj division helps this instability, as opposed to switching to `complex128` slows computation down quite a bit. 

We were additionally confused because on CPU the error is different from the GPU error. On CPU the error that is thrown mentions that the eigenvalues are too close to one another before the `linalg_eig_backward` call, but on GPU we crash at the below call:

 `The eigenvectors in the complex case are specified up to multiplication by e^{i phi}. The specified loss function depends on this quantity, so it is ill-defined`.

If degeneracy of the matrix we take `eigh` with is the issue, then I understand that we would run into the error that we run into on CPU below:

`torch._C._LinAlgError: linalg.eigh: (Batch element 0): The algorithm failed to converge because the input matrix is ill-conditioned or has too many repeated eigenvalues (error code: 7).`

This instability also exists with real values, as with real dtypes (32-bit float) we see some `NaN` gradients produced, however without throwing an error.

Apologies if I am misunderstanding anything at all but wanted to see the team's thoughts on adding the `eps` for numerical stability in any case, or at least raise the issue of different errors between CPU and GPU.

Thank you!

**Comment by alexanderlerner** [comment]
Hi @lezcano , thank you for your response! Apologies for the delayed reply as I wanted to investigate this further before responding.

It is after the check, but in our case the division of `VhgV.div_(Econj)` produces NaNs for the gradient because `Econj` is equal to `0`. Then, in the next iteration, the NaNs that are produced seem to trigger the `TORCH_CHECK` in the subsequent iteration.

We find that adding the `eps` to the Econj division helps this instability, as opposed to switching to `complex128` slows computation down quite a bit.

We were additionally confused because on CPU the error is different from the GPU error. On CPU the error that is thrown mentions that the eigenvalues are too close to one another before the `linalg_eig_backward` call, but on GPU we crash at the below call:

 `The eigenvectors in the complex case are specified up to multiplication by e^{i phi}. The specified loss function depends on this quantity, so it is ill-defined`.

If degeneracy of the matrix we take `eigh` with is the issue, then I understand that we would run into the error that we run into on CPU below:

`torch._C._LinAlgError: linalg.eigh: (Batch element 0): The algorithm failed to converge because the input matrix is ill-conditioned or has too many repeated eigenvalues (error code: 7).`

This instability also exists with real values, as with real dtypes (32-bit float) we see some `NaN` gradients produced, however without throwing an error.

Please let me know if the failure sequence is still unclear or you think I have analyzed incorrectly.  To summarize, we are hoping to get your thoughts on:
- Correcting the error handling to consistently raise the repeated eigenvalues error regardless of device and use of complex types
- Adding the ability for inclusion of an eps term for numerical stability
Thank you!

**Comment by alexanderlerner** [comment]
Thank you @lezcano. 

Agreed that `complex128` can remove that issue, but unfortunately at major performance cost. Also understood on the error discrepancy between CPU and GPU.

We will be using the `eps` argument in `linalg_eig_backward` on our end, but do you have a sense on whether a PR including this additional eps in that division might be accepted? 

We could propose one of the following options:
- Add `eps` only if the value of Econj is below some tolerance value, say `1e-8`
- Add `eps` with magnitude dependent on the dtype of the tensors. Since `complex128` is less "error-prone", we could have smaller magnitude for larger dtypes.
- Add `eps` in a kwarg for the call to eigh as an optional parameter

Please let me know if that is something that you or another contributor would be open to and I'd be happy to add a PR. 

Thanks for all your help!

**Comment by lezcano** [comment]
This issue is already documented, and this is not a formal solution to it, so I don't think we'd accept it.

**Comment by alexanderlerner** [comment]
Understood, thank you
